### PR TITLE
Cleanup leds

### PR DESCRIPTION
**Pull request by ledvinap**
Cleanup of ledstrip code. The changes mostly do not affect old behavior, only code is improved.
Over 500 bytes of FLASH saved:
```
 116984	   1072	  13452	 131508	  201b4	./obj/main/cleanflight_SPARKY_new.elf
 117536    1072   13452  132060   203dc ./obj/main/cleanflight_SPARKY_orig.elf
```

### CODE REVIEW

**Comment by ledvinap** [comment]
Unittests needs update, I'll comit it soon

**Comment by gaelj** [comment]
Wonderful job. With this much flash saved, couldn't you re enable GPS LEDs on the CC3D ?

File: src/test/unit/ledstrip_unittest.cc [hunk 1]
```diff
@@ -151,21 +156,21 @@ TEST(LedStripTest, parseLedStripConfig)
     memset(ledConfigs_arr(), 0, sizeof(*ledConfigs_arr()));
 
     // and
-    bool ok = false;
+    bool ok = true;
```
Comment (gaelj) [inline src/test/unit/ledstrip_unittest.cc]: rename "ok" to "result" ?

Comment (ledvinap) [inline src/test/unit/ledstrip_unittest.cc]: maybe ... but it means 'OK so far', ..

Comment (gaelj) [inline src/test/unit/ledstrip_unittest.cc]: so it's the running result ;-)

**Comment by ledvinap** [comment]
Here is list of questionable parts led code, comments are welcome

Some colors are hardcoded, avoiding user-configurable palette - WARNING, INDICATOR, and partially color ring. Moving them into specialColors is probably way to go.

Default (ROM based) palette is not user accessible. Color index may map both into user-defined colors and hardcoded ones (for example using negative index for hardcoded colors)

Grid size and quadrant handling works only if leds start at index 0,0. When using configurator, it is natural to use grid center as reference, but then quadrant and orientation handling does not work.
It may be better to hardcode north/south/east/west limits (possibly with wide gap in center, so that user can enough place leds that are part of any side). Grid size calculation will be a bit more complicated (min/max bounds / bitmap)

`applyDirectionalModeColor` - I don't understand the logic - why must be led in correct quadrant AND correct direction? Is there some use-case when this is useful ?

WARNING layer - the cycle is fixed length, even when only one warning is active. Skipping inactive warnings (instead of waiting) may be better choice.

`applyLedGpsLayer` - when there is no 3d fix, led alternates between GPSNOLOCK and GPSNOSATS. Is this intended? @gaelj ?

`applyLedThrustRingLayer` - sequence length algorithm is working only in few cases. I'll fix it .. 
What if we rotate actual led colors? Let user select colors + pattern ..

`applyLedAnimationLayer` - Can someone explain this mode?

**Comment by gaelj** [comment]
@ledvinap 
- remaining hard-coded colors: indicator and color ring should indeed be user-configurable. I decided to deliberatrley skip Warning colors to prevent support issues.
- about orange / red blinking, this was indeed not the original behaviour which was orange / black.
@digitalentity introduced it in his proposal here https://gist.github.com/digitalentity/f4f021876653acb1f7f1525024388348 and I personnaly found it better than the old method so I kept it.
- if you enable USE_LED_ANIMATION, you will see what applyLedAnimationLayer does. Basically it adds a circling white LED animation, which is not configurable and in my opinion, annoying (especially when used with the new GPS mode as it acts as a distraction, makes it slightly more difficult to read the satellite count as green LED blinks). It does look pretty cool on first look though.
- an additional point: specialcolors is a 1-element array. I implemented it in this way because it was easier but we could probably optimize this.

**Comment by ledvinap** [comment]
`applyLedIndicatorLayer` lights only leds that are in some quadrant (SE,SW,NE,NW). Leds that are directly N/E/S/W are not affected. Is this desired behavior?

**Comment by gaelj** [comment]
I believe it is, probably due to being designed primarily for X shaped quads.
In my opinion it could be generalised to all horizontal directions, preferably in a user-configurable way.

**Comment by ledvinap** [comment]
There is ugly bug in master ledstrip code

`applyLedWarningLayer`, `applyLedGpsLayer` and `applyLedBlinkLayer` first update led state, then check `updateNow` flag and update internal state.
When other event is triggered before next `updateNow == true` call (which is expected behavior), it will apply internal state to led state, changing the led too soon. This causes too short or too long blinks and possibly blinks missing at all (GPS led problems were probably caused by this bug).

**Comment by ledvinap** [comment]
New version committed, ready for testing.

I'll probably rebase this branch to cleanup commits, so expect force-push.

Code size is reduced by 950B, new code is also a faster (but take that with grain of salt)

```
   text    data     bss     dec     hex filename
 116584    1072   13452  131108   20024 ./obj/main/cleanflight_SPARKY.elf
 117536    1072   13452  132060   203dc ./obj/main/cleanflight_SPARKY.elf

Task list          max/us  avg/us rate/hz maxload avgload     total/ms
13 -     LEDSTRIP     363      34     100    4.1%    0.8%      1538
13 -     LEDSTRIP     242      13      94    2.7%    0.6%      1345
```

**Comment by ledvinap** [comment]
Comparing gyro / blink total time, master is around 1.4%, new version ~ 0.95% - about 35% faster

**Comment by gaelj** [comment]
I will review and test asap (hopefully still today).

**Comment by gaelj** [comment]
Looks all good as far as I can see. Thanks for this cleanup !
@hydra it would be nice to get this merged in 1.14 RC2 as it fixes some bugs in the LED strip handling, cleans up the code, and frees some memory, thus enabling GPS LEDs on CC3D.

**Comment by ledvinap** [comment]
Commits cleaned up, code is identical

Ready for merge

**Comment by hydra** [comment]
great, i will review and test soon.  just got to Malaga in Spain now and set my computers back up again, hence my absence recently.

`applyDirectionalModeColor - I don't understand the logic - why must be led in correct quadrant AND correct direction? Is there some use-case when this is useful ?`

yes.  e.g rear facing LEDs on the front arms.  it's specifically for night-time orientation.  different colors are picked due to the position and led orientation.  

the implementation still allows you to, for example, iterate over all rear-facing leds and do a color fade front to back on rear facing LEDs by row if so desired.

**Comment by hydra** [comment]
ok i gave this a whirl today but there seems to be an issue with the blink color.  it seems the blink color inherits the color from the previous led in the strip.

e.g.

led 0 white,
led 1 blink.

```
led 0 7,8::C:1
led 1 8,8::B:0
```

first led appears white,
second led is white and blinks to OFF.

led 0 white
led 2 red + blink.

```
led 0 7,8::C:1
led 1 8,8::CB:2
```

first led appears white.
second is white and blinks red.

i think the 'bleeding through from previous LED color' is a mistake.

**Comment by hydra** [comment]
documentation says:

```
#### Blink

This mode blinks the current LED, alternatively from black to the selected color.
```

**Comment by hydra** [comment]
a configurable color for the default state for LEDs that blink is probably the right long-term answer here.

**Comment by gaelj** [comment]
Configurable background blink is already the way it's done  : 
`const hsvColor_t *blinkColor = ledOn ? colors(ledConfig->color) : getSC(LED_SCOLOR_BLINKBACKGROUND);`

You can set it via cli command `mode_color 6 4 [colorIndex]`.

I tried your settings and unfortunately I can't reproduce your problem.

**Comment by ledvinap** [comment]
@hydra: It may be better to apply led layers explicitly - using multiple function records for single led instead of flags. With proper encoding, this could save lot of space. 
Configurator can translate from current configuration user-friendly style (with @gaelj enhancements) to this machine-optimized format .. 

Something like:
```
led0: color(blue);
led1: indicator(left); blink(red);
led3: color(green); throttle(); warning();
led4: thrustRing(pos=1);
...
```

**Comment by hydra** [comment]
@gaelj thanks for the update. it looks like the board i was testing with had some garbage in the mode color configuration record as follows:

```
mode_color 6 4 42
mode_color 6 5 0
mode_color 6 6 45
mode_color 6 7 0
```

after resetting to defaults I confirm it's OK now.

**Comment by hydra** [comment]
@ledvinap that's a *really* good refactoring.  nice work!

**Comment by hydra** [comment]
merged, too much good stuff here to leave, further testing any more bug fixing can be done during RC2 phase.

**Comment by gaelj** [comment]
@hydra, would it be too late to merge the LED Battery and MSP mode_color PRs I issued, based on this PR ? I know they are both new features rather than bugfixes, so I'm afraid I already know the answer. :-)
MSP mode_color is required to fuly take advantage of the CFC LED tab rework, and Battery color is a very nice additional feature.
I just merged these 2 PR's with the latest master, rebased and force-pushed.
#2160 
#2173

**Comment by ledvinap** [comment]
@gaelj : I would postpone #2160 a bit, so there is time to find better alternative to expanding size of led config ...

**Comment by hydra** [comment]
@gaelj yes, a bit late.  since I already tagged the RC2 release :smile:

**Comment by gaelj** [comment]
No problem, I made sure the CFC LED tab rework was "retro"-compatible with MW API 1.18.0.

**Pull request merged**